File: 72-EditDistance.py
class Solution:
    # @param {string} word1
    # @param {string} word2
    # @return {integer}
    def minDistance(self, word1, word2):
        word1_len = len(word1) + 1 
        word2_len = len(word2) + 1
        dp = [[0 for j in xrange(word1_len)] for i in xrange(word2_len)]
        for i in xrange(word2_len):
            dp[i][0] = i
        for j in xrange(word1_len):
            dp[0][j] = j
        for i in xrange(1, word2_len):
            for j in xrange(1, word1_len):
                dp[i][j] = min(dp[i][j -1] + 1, dp[i - 1][j] + 1, dp[i - 1][j - 1] + [0, 1][word2[i - 1] != word1[j -1]])
        return dp[-1][-1]

if __name__ == '__main__':
   s = Solution()
   print s.minDistance('e', 'g')
   print s.minDistance('egg', 'dog')
   print s.minDistance('a', 'ab')

File: C++/72-editDIstance.cpp
#include "include/common.h"

class Solution {
public:
    int minDistance(string word1, string word2) {
        int length1 = word1.length() + 1, length2 = word2.length() + 1, one;
        vector<vector<int> > editCounts(length1, vector<int>(length2, 0));
        for (int i = 0; i < length1; i++) {
             editCounts[i][0] = i;
        };
        for (int j = 0; j < length2; j++) {
             editCounts[0][j] = j;
        };
        for (int i = 1; i < length1; i++) {
            for (int j = 1; j < length2; j++) {  
                one = word1[i - 1] != word2[j - 1];
                editCounts[i][j] = min(editCounts[i - 1][j - 1] + one, min(editCounts[i - 1][j] + 1, editCounts[i][j - 1] + 1));
            };
        };
        return editCounts[length1 - 1][length2 - 1];      
    }
};

int main(void) {
    Solution s;
    cout << s.minDistance("mart", "karma") << endl;
    return 0;
};